### ISSUE TITLE
[AdminList] It is not possible to sort entites by group

### ISSUE THREAD

**Issue by mtnorthrop** [Kunstmaan/KunstmaanBundlesCMS#1066]
It's currently not possible to sort entities within a group.

When a new entity is added to an admin list through the `AdminListController::doAddAction()` method, `AdminListController::getMaxSortableField()` is called and returns the highest sorting value, without any option for grouping the entities according to a category or a foreign key.

In my case, I need to sort entities, grouped by foreign key. Since the built-in sortable functionality can not do this, I'm using the Gedmo Sortable Doctrine extension instead.

Overriding the default `moveUp()` and `moveDown()` methods to work with Gedmo Sortable wasn't a problem, but the `doAddAction()` method calls `AdminListController::getMaxSortableField()`, which is a private.

In order to get around this limitation I've overridden the whole `AdminListController::doAdminAction()` method, but I don't feel like this is an ideal solution.

Would it be possible to allow developers to implement their own method for setting the weight value for new entities? Maybe by moving the following code to its own protected method that can be overridden?

```
[From AdminListController::doAddAction()]

// Check if Sortable interface is implemented
if ($configurator instanceof SortableInterface) {
    $sort = $configurator->getSortableField();
    $weight = $this->getMaxSortableField($repo, $sort);
    $setter = "set".ucfirst($sort);
    $helper->$setter($weight + 1);
}
```

This way, if a developer wants to use the Gedmo Sortable extension, they can simply override the parent method with a blank method.

There's probably a much better and more systematic solution, but this would solve the immediate problem.

**Comment by sandergo90** [comment]
Hi,

Do you just want to sort the admin list ? Because that way you can just adapt the query builder and change the orderby clause.

Sander

**Comment by mtnorthrop** [comment]
This is actually a problem when trying to add a new entity to an admin list, not so much when fetching them for the view.

For example, let's say I have a Car entity, but I want to have a separate admin list for each Brand. It's no problem to create an admin list for this by adapting the query builder; **however**, when I create a new Car of a certain Brand using the `AdminListController::doAddAction()` method, it uses the private `AdminListController::getMaxSortableField()` method, which gets the highest sortable value from the database and sets the new entity's sortable value as one more. 

I think that, when getting the highest sortable value, it should be possible add conditions. For example, to get the highest sortable value from the Cars table, only counting rows with a brand_id of "5". 

Basically, I want to be able to do what Gedmo Sortable is doing with the $position property in the example here: https://github.com/Atlantic18/DoctrineExtensions/blob/master/doc/sortable.md#sortable-annotations

**Comment by mlebkowski** [comment]
true, @mtnorthrop, I had the same issue and ended up overriding the `moveUp`/`moveDown` methods as well. I can’t remember, but I guess new items wasn’t an issue for me.

In my opinion this should be refactored, best if the `AdminListConfigurator` could return an adapted query builder and the `getMaxSortableField` method (or a new service extracted from it) use it instead of a fresh query builder. This way you shouldn’t have to override anything in your application code.

**Comment by mtnorthrop** [comment]
I solved the issue by overriding AdminListController methods also. Personally, I'd suggest using the Gedmo DoctrineExtensions Sortable behavior for all of the AdminList sorting functionality, especially since KunstmaanBundles already depends on the Gedmo DoctrineExtensions package anyway.

**Comment by jockri** [comment]
@mtnorthrop can you send a PR to make it more easy to override by moving some code to a protected function?

**Comment by mtnorthrop** [comment]
I'd be happy to do this, but I won't be able to right away. I'll submit a PR as soon as I can, though.

**Comment by mtnorthrop** [comment]
The new method that was added in the merge fixes this issue, so I'm closing it.
